Match pattern fragment clarification (#36069)
mdn/content: files/en-us/mozilla/add-ons/webextensions/match_patterns/index.md

      <td>Any host.</td>
    </tr>
    <tr>
      <td><code>*.</code> followed by part of the hostname, optionally, including a port.</td>
      <td>The given host (and port) and any of its subdomains.</td>
    </tr>
    <tr>
      <td>A complete hostname, without wildcards, optionally, including a port.</td>
      <td>Only the host (and port).</td>
    </tr>
  </tbody>
</table>

> [!NOTE]

> > Firefox doesn't support the inclusion of a port number due to ([Firefox bug 1362809](https://bugzil.la/1362809)) and ([Firefox bug 1468162](https://bugzil.la/1468162)).

Author: @dotproto re the comments about port numbers in https://github.com/mdn/content/pull/36069/files#r1986494273, doesn't this address that issue?

      <td><code>*.</code> followed by part of the hostname.</td>
      <td>The given host and any of its subdomains.</td>
    </tr>
    <tr>

>       <td><code>*.</code> followed by part of the hostnamp plus a port number.</td>

Reviewer: The port is optional.

Author: Hence, why this follows <td><code>*.</code> followed by part of the hostname.</td>

Reviewer: Ah! My mistake. I missed the context of content outside the diff view.

The value for the _path_ matches against the string which is the URL path plus the [URL query string](https://en.wikipedia.org/wiki/Query_string). This includes the `?` between the two, if the query string is present in the URL. For example, if you want to match URLs on any domain where the URL path ends with `foo.bar`, then you need to use an array of Match Patterns like `["*://*/*foo.bar", "*://*/*foo.bar?*"]`. The `?*` is needed, rather than just `bar*`, in order to anchor the ending `*` as applying to the URL query string and not some portion of the URL path.

> 

Author: ```suggestion

Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` which precedes it are considered as part of the _path_. This means that fragments aren't considered when evaluating match patterns.
```

Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier), nor the `#` which precedes it, are considered as part of the _path_.

> [!NOTE]

> > The path pattern string should not include a port number. Adding a port, as in: `http://localhost:1234/*` causes the match pattern to be ignored. However, `http://localhost:1234` will match with `http://localhost/*`.

Author: @dotproto @Rob--W is this still relevant, given host can include ports? (Or should it be rephrased more along the lines of "if you include a port in the host pattern, you can't include a path.")

Reviewer: I think it's still relevant to say that fragments aren't considered when evaluating match patterns. IMO we should also recommend omitting them.

The note portion isn't accurate. There are some situations where an extension may explicitly want to connect to a specific port. For example, an application might expose a web server on `localhost` using a well-known port, or a web server may be publicly accessible on the Internet using a non-standard port. In both cases specify a port is necessary. 

If we were to retain the note, I'd suggest tweaking it to recommend against including the port number unless it is strictly necessary. For example, prefer `http://example.com/*` over `http://example.com:80/*`. Over-specifying the host in a match pattern may result in unexpected or inconsistent behavior. For example, `http://example.com:443/*` will not match when a user navigates to `http://example.com/` in Firefox, but it will in Chrome.

Author: > I think it's still relevant to say that fragments aren't considered when evaluating match patterns. IMO we should also recommend omitting them.

@dotproto is this addressed by the proposed change above?

The value for the _path_ matches against the string which is the URL path plus the [URL query string](https://en.wikipedia.org/wiki/Query_string). This includes the `?` between the two, if the query string is present in the URL. For example, if you want to match URLs on any domain where the URL path ends with `foo.bar`, then you need to use an array of Match Patterns like `["*://*/*foo.bar", "*://*/*foo.bar?*"]`. The `?*` is needed, rather than just `bar*`, in order to anchor the ending `*` as applying to the URL query string and not some portion of the URL path.

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Author: Fragments in the URL are ignored - first sentence -  (therefore) a match pattern containing fragments fails to match any URL - second sentence.
```suggestion
Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. Therefore, a match pattern containing `#` will fail to match with any URL.
```

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Author: Ping @dotproto

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Reviewer: Ah, I missed your last comment because I was only looking at my pull request queue. 

Looks like we may have different views on what "match" means in this context. I don't think it's accurate to say "a match pattern containing `#` will fail to match with any URL." 

As I see it, if an extension requests access to a given pattern, then any URL on which the extension receives additional access grants as a result is a match for that pattern. The actual string the developer typed or their intent when typing it don't affect whether or not a pattern is matched.

More concretely, if I declare `"http://example.com/sub/path/#my-fragment"` in `host_permissions` an extension, install it, and then visit `http://example.com/`, the extension will be able to access the URL of that tab. The extension will not match URLs with a fragment as I may have wanted, but the pattern I declared can still match URLs that I visit.

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Author: @dotproto However, isn't your example incorrect in that "http://example.com/sub/path/#my-fragment" would only match with "http://example.com/sub/path/"? Presumably, the fragment is not considered part of the path in either the match pattern or the URL being matched. Hence:
```suggestion
Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered part of the _path_ and are ignored during pattern matching. Therefore, a match pattern containing `#` will fail to match with only the URL containing the fragment. For example, in the match pattern `"http://example.com/sub/path/#my-fragment"` the browser will look for a match on `"http://example.com/sub/path/"`. This matches with any URL containing that host and path and any fragment. So, it matches `"http://example.com/sub/path/#my-fragment"`,  `"http://example.com/sub/path/#another-fragment"`, etc.
```

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Reviewer: @rebloor @Rob--W, in order to make sure we're all on the same page, I put together an example extension that tests a bunch of host permissions match patterns and how various URLs do (not) match them. https://gist.github.com/dotproto/f199f3f0ed96b50761e4b537c72f26b8

As you can see in the tests for `"http://www.example.edu/test/*#fragment"`, Chrome and Firefox both ignore the entire path for host permissions checks. This does not apply to other types of match patterns, such as the `matches` field in a content script declaration.

> Neither the [URL fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) nor the `#` that precedes it are considered as part of the _path_ and are ignored during pattern matching. A match pattern containing `#` will fail to match with any URL.

Reviewer: I started looking at how Safari handles host permissions declarations and it's a whole different world. It's so different that it simply has to be out of scope for the time being.

  </tbody>
</table>

### Invalid or unmatched patterns

<table class="fullwidth-table standard-table">
  <thead>
    <tr>
      <th scope="col">Pattern</th>
      <th scope="col">Issue</th>
      <th scope="col">Reason</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>resource://path/</code></td>
      <td>Invalid</td>
      <td>Unsupported scheme.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.org</code></td>
      <td>Invalid</td>
      <td>No path.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.org/#section1</code></td>
      <td>Unmatched</td>
      <td>Contains a reference fragment: the URL that the pattern is matched against has any reference fragment removed before matching.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.*.org/</code></td>
      <td>Unmatched</td>

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Author: FYI @Rob--W does this mean I misunderstood your comment? ("On the other hand, the example below with a host name is also syntactically valid but listed as invalid. In Chrome it is valid but in Firefox it is ignored.")

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Author: @dotproto also commented on the port number example [here](https://github.com/mdn/content/pull/36069#discussion_r1849497036), pointing out that there was a mismatch between the protocol and port number. So, to clarify, the statement "host must not include a port number." is incorrect? Therefore, we should:

- add a port to the details of how the [host](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#host) can be specified with a note about the issues in Firefox
- add an example to the [examples](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#examples) with a note that the port number must match the protocol
- remove the port example from the invalid list

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Author: @dotproto @Rob--W Any feedback on my comments?

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Reviewer 2: > - add a port to the details of how the [host](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#host) can be specified with a note about the issues in Firefox

Looks like there is already a note "host must not include a port number.". This can be revised to limit to Firefox and include a link to the Firefox bug.

(I just checked [WebKit's source code](https://searchfox.org/wubkat/rev/795fdaece398b18bdefcd3954e0da2312ca92ee9/Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.cpp#409) and confirmed that Safari supports ports too)

> - add an example to the [examples](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#examples) with a note that the port number must match the protocol

I don't know why you would suggest that. The default port for http is 80, for https 443. When a user or website navigates there, the port is automatically omitted because it is the default port. So if a match pattern were to look like `https://example.com:443/*`, then it would never match because extensions would never encounter such a URL in practice.

Your suggestion would basically be a "bad example" example. It is a syntactically valid match pattern, but known to never match anything.

> - remove the port example from the invalid list

Could move the example to the good section with a remark that it does not work in Firefox + bug number.

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Author: @Rob--W 

>  * add an example to the [examples](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#examples) with a note that the port number must match the protocol
> 
> I don't know why you would suggest that. The default port for http is 80, for https 443. When a user or website navigates there, the port is automatically omitted because it is the default port. So if a match pattern were to look like `https://example.com:443/*`, then it would never match because extensions would never encounter such a URL in practice.

Because of @dotproto [comment](https://github.com/mdn/content/pull/36069#discussion_r1849497036):

> Ugh, we're in weird territory now. This pattern (https://mozilla.org:80/) doesn't work for two reasons: the aforementioned origin redirect and a mismatch between the declared port and port required by the scheme. HTTPS runs on port 443, HTTP runs on port 80.

> Neither Chrome or Firefox will log an install time error/warning if an extension requests https://mozilla.org:80/. This permission is "valid" (but meaningless). You can see this in action if you use await chrome.permissions.getAll() to check the extension's current permission; the response object will look something like this: { permissions: [], origins: ["https://mozilla.org:80/*"]}.

> If you modify the host permission to request a port that matches the scheme (e.g. https://www.mozilla.org:443/ or http://example.com:80/), both Chrome and Firefox will grant the host permission and the extension will be able to access data for that host.

So essentially, because the "bad" example is `https://mozilla.org:80/` so includes the wrong port for the protocol, but doesn't say that so. And I was suggesting adding a "good" example with the note.

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Reviewer 2: Simeon's comment and mine are not contradicting.

The current example is `https://mozilla.org:80/`. The original explanation was `Host must not include a port number.`. Simeon states that while the example does not work in Chrome, it failed for a different reason than in Firefox. Firefox does simply not support ports, whereas in Chrome the match pattern is not going to match anything, in practice. HTTPS runs on port 443, HTTP runs on port 80. In theory a website could try serving HTTPS over port 80 or HTTP over port 443, but let's ignore that for now.

I can confirm that Chrome supports match patterns where the scheme is 443 (https) or 80 (http). They still do not work in Firefox because ports are not supported.

In any case, the message is not "Port number must match the protocol.", but "Ports are supported in Chrome, not in Firefox", and as an example you could use port number 8080 for example.

>       <td>Is syntactically valid and matches in Chrome but is ignored in Firefox.</td>

Author: Thanks @Rob--W. However, it doesn't make sense to include the port example in the invalid examples list so I've moved it to the valid examples list with the note about support.

    </tr>
    <tr>
      <td><code>https://*zilla.org/</code></td>
      <td>Invalid</td>
      <td>"*" in host must be the only character or be followed by ".".</td>
    </tr>
    <tr>
      <td><code>http*://mozilla.org/</code></td>
      <td>Invalid</td>
      <td>"*" in scheme must be the only character.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.org:80/</code></td>

>       <td>Invalid</td>

Reviewer: Ugh, we're in weird territory now. This pattern (`https://mozilla.org:80/`) doesn't work for two reasons: the aforementioned origin redirect and a mismatch between the declared port and port required by the scheme. HTTPS runs on port 443, HTTP runs on port 80. 

Nether Chrome or Firefox will log an install time error/warning if an extension requests `https://mozilla.org:80/`. This permission is "valid" (but meaningless). You can see this in action if you use `await chrome.permissions.getAll()` to check the extension's current permission; the response object will look something like this: `{ permissions: [], origins: ["https://mozilla.org:80/*"]}`.

If you modify the host permission to request a port that matches the scheme (e.g. `https://www.mozilla.org:443/` or `http://example.com:80/`), both Chrome and Firefox will grant the host permission and the extension will be able to access data for that host.

Author: @dotproto so should we use another port, such as 8080. I assume from your comment that this was just a poorly chosen example given that 80 has a binding, or is.the statement "_host_ must not include a port number." Incorrect

      <td>Host must not include a port number.</td>
    </tr>
    <tr>
      <td><code>*://*</code></td>
      <td>Invalid</td>
      <td>Empty path: this should be "<code>*://*/*</code>".</td>
    </tr>
    <tr>
      <td><code>file://*</code></td>

>       <td>Invalid</td>

Reviewer: `file://*` is invalid when declared in `host_permissions` in Firefox but works in Chrome. Chrome auto-corrects it to `file:///*`. 

`file:///*` works in both Firefox and Chrome.

Author: Added note re-Chrome behavior

  </tbody>
</table>

### Invalid or unmatched patterns

<table class="fullwidth-table standard-table">
  <thead>
    <tr>
      <th scope="col">Pattern</th>
      <th scope="col">Issue</th>
      <th scope="col">Reason</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>resource://path/</code></td>
      <td>Invalid</td>
      <td>Unsupported scheme.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.org</code></td>
      <td>Invalid</td>
      <td>No path.</td>
    </tr>
    <tr>
      <td><code>https://www.mozilla.org/#section1</code></td>
      <td>Unmatched</td>

>       <td>Contains a reference fragment: the URL that the pattern is matched against has any reference fragment removed before matching.</td>

Author: As I understand it, that's not the point. The point is that if you have a URL of `https://www.mozilla.org/#section1`, as per the text in the previous comment, it has the fragment removed before matching. So the URL used to match against the match pattern is `https://www.mozilla.org/` (the match pattern is unchanged).

      <td><code>https://mozilla.org</code></td>
      <td>No path.</td>
    </tr>
    <tr>
      <td><code>https://mozilla.org/#section1</code></td>

>       <td>Contains an unsupported match character.</td>

Reviewer 2: This is not a syntactically invalid match pattern, just one that never matches.

The other examples here are examples of match patterns that are rejected as a value for "match pattern" types, but this one is accepted.

On the other hand, the example below with a host name is also syntactically valid but listed as invalid. In Chrome it is valid but in Firefox it is ignored.

Reviewer: The basic structure of this pattern does match when used in `host_permissions`. I tested this using a modified version of [this test extension](https://gist.github.com/dotproto/f199f3f0ed96b50761e4b537c72f26b8) and received the following output from Chrome and Firefox.

Firefox 133.0 buildID 20241121140525 
```text
manifest patterns (1): ["http://example.com/#fragment"]
granted  patterns (1): ["http://example.com/*"]
pattern  origins  (1): ["http://example.com"]
  Request URL                                        Observed URL (indicates pattern match)
Testing pattern "http://example.com/#fragment"
  http://example.com                                 tab.url = http://example.com/
  http://example.com#fragment                        tab.url = http://example.com/#fragment
  http://example.com/                                tab.url = http://example.com/
  http://example.com/#fragment                       tab.url = http://example.com/#fragment
  http://example.com/subpath                         tab.url = http://example.com/subpath
  http://example.com/subpath#fragment                tab.url = http://example.com/subpath#fragment
  http://example.com/sub/path/                       tab.url = http://example.com/sub/path/
  http://example.com/sub/path/#fragment              tab.url = http://example.com/sub/path/#fragment
```

Chrome Version 131.0.6778.86 (Official Build) (arm64)
```text
manifest patterns (1): ["http://example.com/#fragment"]
granted  patterns (1): ["http://example.com/*"]
pattern  origins  (1): ["http://example.com"]
  Request URL                                        Observed URL (indicates pattern match)
Testing pattern "http://example.com/#fragment"
  http://example.com                                 tab.url = http://example.com/
  http://example.com#fragment                        tab.url = http://example.com/#fragment
  http://example.com/                                tab.url = http://example.com/
  http://example.com/#fragment                       tab.url = http://example.com/#fragment
  http://example.com/subpath                         tab.url = http://example.com/subpath
  http://example.com/subpath#fragment                tab.url = http://example.com/subpath#fragment
  http://example.com/sub/path/                       tab.url = http://example.com/sub/path/
  http://example.com/sub/path/#fragment              tab.url = http://example.com/sub/path/#fragment
```

Safari Version 18.1 (20619.2.8.11.10)
```text
manifest patterns (1): ["http://example.com/#fragment"]
granted  patterns (1): ["*://*.example.com/*"]
pattern  origins  (1): ["http://example.com"]
  Request URL                                        Observed URL (indicates pattern match)
Testing pattern "http://example.com/#fragment"
  http://example.com                                 tab.url = http://example.com/
  http://example.com#fragment                        tab.url = http://example.com/#fragment
  http://example.com/                                tab.url = http://example.com/
  http://example.com/#fragment                       tab.url = http://example.com/#fragment
  http://example.com/subpath                         tab.url = http://example.com/subpath
  http://example.com/subpath#fragment                tab.url = http://example.com/subpath#fragment
  http://example.com/sub/path/                       tab.url = http://example.com/sub/path/
  http://example.com/sub/path/#fragment              tab.url = http://example.com/sub/path/#fragment
```

>       <td>Contains an unsupported match character.</td>

Reviewer 2: ```suggestion
      <td>Contains a reference fragment, which are never included in the URLs to match against.</td>
```

Can probably be phrased better than this. It is not that the character is unsupported technically, but that the URL that the pattern is matched against never includes the reference fragment.

Author: @Rob--W made some changes to address, let me know what you think
